### PR TITLE
Expr2

### PR DESCRIPTION
**Pull request by Dieterbe**
note: as you can tell from the NOTES, i'm not really happy with some of the current ugliness regarding parse-time objects making their way into the planner and function execution. i want to refactor that. but it'll be quite the job, and this seems good enough for now.

### CODE REVIEW

File: expr/NOTES [hunk 1]
```diff
@@ -1,3 +1,21 @@
...
+function invocations could be attached to expressions (when creating the plan). when executing them, we then
+also don't need to pass args anymore (which is currently redundant since they already got them via Init)
+* in general, while constructing the plan, we should probably build a different datastructure more tailored towards execution, currently we
+  stick to the parse-time expr types in the planner and these concepts also make their way into the function implementations, which gets a bit dirty at times.
+
+
+## regarding when to allocate models.Series (or more precisely, their []schema.Point attribute)
+
+there's 2 main choices
 regarding when to allocate models.Series (or more precisely, their []schema.Point attribute), there's 2 main choices:
```
Comment (woodsaj) [inline expr/NOTES]: looks like this should be deleted.

Comment (Dieterbe) [inline expr/NOTES]: why? I like to keep record of what the reasons were behind the implementation.

Comment (woodsaj) [inline expr/NOTES]: the line that should be deleted is duplicate content of the 2 lines above it.

**Comment by woodsaj** [comment]
The planning phase feels a little convoluted.  Arguments are being validated in newplanFunc() and also in each graphiteFunctions Init() method.


There are also a couple of issues with this implementation.
1) maxDatapoints and consolidateBy dont seem to be used. So we are likely sending more data to user then we should be
2) all targets are planned and executed together.  once we start consolidating down to maxDatapoints, a consolidateBy in one target will affect he consolidateBy in another.


i have created a new branch that refactors this one to simplify the logic with execution planning.
https://github.com/raintank/metrictank/pull/625

**Comment by Dieterbe** [comment]
see also #630 re consolidation

**Pull request closed without merging**